rkt/api_service.go

```diff
@@ -0,0 +1,557 @@
+// The checkFunc takes two strings, and returns whether the two strings satisfy the
+// given condition.
+func findString(needle string, haystack []string, checkFunc func(a, b string) bool) bool {
+	for _, v := range haystack {
+		if checkFunc(needle, v) {
+			return true
+		}
+	}
+	return false
+}
+
+// stringEqual returns true if two strings are equal.
+func stringEqual(a, b string) bool {
+	return a == b
+}
+
+// hasBaseName returns true if the second string is the base name of
+// the first string.
+func hasBaseName(name, baseName string) bool {
+	return path.Base(name) == baseName
+}
+
+// findAnyString returns true if there's any two-string pair from array a and
+// array b that satisfy the checkFunc.
+//
+// e.g. if a = {"a", "b", "c"}, b = {"c", "d", "e"} and c = {"e", "f", "g"},
+// then findAnyString(a, b, stringEqual) == true,
+//      findAnyString(a, c, stringEqual) == false,
+//      findAnystring(b, c, stringEqual) == true.
+//
+func findAnyString(a, b []string, checkFunc func(a, b string) bool) bool {
+	for _, aa := range a {
+		if findString(aa, b, checkFunc) {
+			return true
+		}
+	}
+	return false
+}
+
+// filterPod returns true if the pod doesn't satisfy the filter, which means
+// it should be filtered and not be returned.
+// It returns false if the filter is nil or the pod satisfies the filter, which
+// means it should be returned.
+func filterPod(pod *v1alpha.Pod, manifest *schema.PodManifest, filter *v1alpha.PodFilter) bool {
+	// No filters, return directly.
+	if filter == nil {
+		return false
+	}
+
+	// Filter according to the state.
+	if len(filter.States) > 0 {
+		foundState := false
+		for _, state := range filter.States {
+			if pod.State == state {
+				foundState = true
+				break
+			}
+		}
+		if !foundState {
+			return true
+		}
+	}
+
+	// Filter according to the app names.
+	if len(filter.AppNames) > 0 {
+		var names []string
+		for _, app := range pod.Apps {
+			names = append(names, app.Name)
+		}
+		if !findAnyString(names, filter.AppNames, stringEqual) {
+			return true
+		}
+	}
+
+	// Filter according to the image IDs.
+	if len(filter.ImageIds) > 0 {
+		var ids []string
+		for _, app := range pod.Apps {
+			ids = append(ids, app.Image.Id)
+		}
+		if !findAnyString(ids, filter.ImageIds, stringEqual) {
+			return true
+		}
+	}
+
+	// Filter according to the network names.
+	if len(filter.NetworkNames) > 0 {
+		var names []string
+		for _, network := range pod.Networks {
+			names = append(names, network.Name)
+		}
+		if !findAnyString(names, filter.NetworkNames, stringEqual) {
+			return true
+		}
+	}
+
+	// Filter according to the annotations.
+	if len(filter.Annotations) > 0 {
+		if !findKeyValue(manifest.Annotations, filter.Annotations) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getBasicPod returns *v1alpha.Pod with basic pod information, it also returns a *schema.PodManifest
+// object.
+func getBasicPod(p *pod) (*v1alpha.Pod, *schema.PodManifest, error) {
```

Comment:
This function is a bit too long. Wouldn't you mind splitting it into smaller chunks? Like `getBasicPod` could call `getManifest`, then `getPodState`, `getPodPid` and so on.